Implement an optional /config endpoint that outputs the used configuration.

fedepaol opened the pull request:
<!-- 
    Welcome, Athenian! Can you do us two quick favors before you submit your PR?
    
    1. Briefly fill out the sections below. It will make it easy for us to review your code
    2. Put "[WIP]" at the beginning of your PR title if you're not ready to have this merged yet (we have a bot that will tell everyone that it's a work in progress)
-->

## What is the problem I am trying to address?

By setting a ConfigPort parameter, athens will expose a /config endpoint returning the
configuration currently used.
This is useful for debugging purposes. The port is different as the configuration
may contain sensitive informations, and we want to make sure it's exposed only locally.

## How is the fix applied?

Expose the configfile  via a /config endpoint.

## What GitHub issue(s) does this PR fix or close?

<!--
    If it doesn't fix any GitHub Issues, that's ok. Can you please delete the below "Fixes #" line for us? It would help us out a lot. Thanks!

    Your PR might fix one or more GitHub issues. If so, please use the below "Fixes #<issue number>" notation below. If your PR fixes multiple issues, please put multiple lines of "Fixes #<issue number>", one for each issue. If you do that, when this PR is merged, it'll automatically close the issue(s) you reference.
-->

Fixes #1503 

<!-- 
example: Fixes #123
-->

arschles commented on pkg/config/gcp.go:
```diff
@@ -2,7 +2,7 @@ package config
-	ProjectID string `envconfig:"GOOGLE_CLOUD_PROJECT"`
-	Bucket    string `validate:"required" envconfig:"ATHENS_STORAGE_GCP_BUCKET"`
-	JSONKey   string `envconfig:"ATHENS_STORAGE_GCP_JSON_KEY"`
+	ProjectID string `envconfig:"GOOGLE_CLOUD_PROJECT" json:"projectID"`
```
@fedepaol does this affect any of the config decoding we do? Guessing no because it's TOML but just want to check

fedepaol replied:
I assumed it doesn't, as the annotation is for json. Also, looking [at the lib we use](https://github.com/BurntSushi/toml), the annotation syntax is `toml:foo`.

fedepaol replied:
also also, if this was affecting decoding, tests would've failed

arschles commented on pkg/config/s3.go:
```diff
@@ -2,14 +2,14 @@ package config
-	Endpoint                           string `envconfig:"AWS_ENDPOINT"`
+	Region                             string `validate:"required" envconfig:"AWS_REGION" json:"region"`
+	Key                                string `envconfig:"AWS_ACCESS_KEY_ID" json:"key"`
+	Secret                             string `envconfig:"AWS_SECRET_ACCESS_KEY" json:"secret"`
```
@fedepaol I don't think this config value should be returned. And the same for other secrets. Maybe have a struct tag that indicates it should be hidden, and then read that when you encode it for JSON?

fedepaol replied:
As part of the discussion in [https://github.com/gomods/athens/issues/1503#issuecomment-604677096](https://github.com/gomods/athens/issues/1503#issuecomment-604677096) we agreed that it was fine to expose (also) the sensitive data on a different port, as the endpoint was reachable only with direct machine / container access.

twexler commented:
Could you explain how this makes debugging easier? I'm a little weary on adding something that could expose credentials.

fedepaol commented:
> Could you explain how this makes debugging easier? I'm a little weary on adding something that could expose credentials.

@twexler the rationale is describe in the related issue https://github.com/gomods/athens/issues/1503.
The key to safety is having the endpoint exposed under a different point, as discussed with @marwan-at-work  in https://github.com/gomods/athens/issues/1503#issuecomment-604677096 
Hope I did not misunderstand it.

marwan-at-work commented on pkg/config/handler.go:
```diff
@@ -0,0 +1,17 @@
+)
+
+// Handler returns the current configuration as json when hitting the /config url
+func Handler(c *Config) http.HandlerFunc {
```
Let's keep `pkg/config` non-http-aware, and move this file to `cmd/proxy/actions`. We can call it `actions.ConfigHandler`. 

This helps keep all of our handler in one place (of course with exception to the download.Protocol handlers since they are tightly coupled with the logic of a GOPROXY)

marwan-at-work commented on pkg/config/config.go:
```diff
@@ -21,45 +21,46 @@ const defaultConfigFile = "athens.toml"
-	SingleFlight      *SingleFlight
-	Storage           *Storage
-	Index             *Index
+	GoEnv             string        `validate:"required" envconfig:"GO_ENV" json:"goenv"`
```
I don't think we need to add any `json:""` tags in this struct. We can just keep the outputted JSON with upper cases. It's quite common for example the GOPROXY itself outputs is `.info` file with upper case fields: https://proxy.golang.org/github.com/pkg/errors/@v/v0.9.1.info

marwan-at-work commented on pkg/config/config.go:
```diff
@@ -21,45 +21,46 @@ const defaultConfigFile = "athens.toml"
+	SingleFlight      *SingleFlight `json:"singleFlight"`
+	Storage           *Storage      `json:"storage"`
+	Index             *Index        `json:"index"`
+	ConfigPort        *string       `envconfig:"ATHENS_CONFIG_PORT" json:"configPort"`
```
No need to make this a pointer to a string, we can just make it a string and check in main.go for `!= ""` instead of `!= nil`. 

It's pretty common and we do it a lot for the other config fields :v:

marwan-at-work commented on pkg/config/handler_test.go:
```diff
@@ -0,0 +1,114 @@
+	if err := json.Unmarshal(rr.Body.Bytes(), toCheck); err != nil {
+		t.Error("Failed to unmarshal the reply")
+	}
+	if !reflect.DeepEqual(toCheck, testConfig) {
+		t.Error("Returned config different from original")
+	}
```
Let's use `cmp` to check for equality, but more importantly to output a friendly Diff so that if the test fails a user knows why it failed. 

[See this line of code for an example(https://github.com/gomods/athens/blob/23672f78c933b1d7ad4b3353ce628ccfb1b988b1/pkg/config/config_test.go#L28:L32)

fedepaol replied:
Did not know about it, awesome!

marwan-at-work commented on config.dev.toml:
```diff
@@ -99,6 +99,10 @@ EnablePprof = false
 PprofPort = ":3001"
 
+# ConfigPort specifies the port on which the /config endpoint will listen for
+# providing the internal configuration
```
Let's add a note about secrets: "Note if you have secrets in your configuration, this endpoint will return them in plaintext. This is why they are exposed in a different port and should be used for debugging purposes only".

marwan-at-work commented on config.dev.toml:
```diff
@@ -99,6 +99,10 @@ EnablePprof = false
 
+# ConfigPort specifies the port on which the /config endpoint will listen for
+# providing the internal configuration
+#ConfigPort = ":3002"
```
Per my comment about making `*string` into a `string`. You can uncomment this line and just do `ConfigPort = ""` similar to a lot of our configuration fields

marwan-at-work commented:
@twexler  @arschles I hear your point about the security concerns. My idea was that by making this endpoint off-by-default and by forcing the Athens admin to expose it on a different port, that would help mitigate credentials exposure. 

That said, I'm also pretty happy with the alternative: expose the config on the same port but obscure all credentials. 

Happy either way :v:

fedepaol commented:
@marwan-at-work I think I addressed all the comments.

Also, @arschles @twexler let me know if you prefer me to take the other route, and not publish (or obscure) the "sensitive" configuration parameters.

twexler commented:
> Also, @arschles @twexler let me know if you prefer me to take the other route, and not publish (or obscure) the "sensitive" configuration parameters.

I'm still struggling to ascertain the usefulness of this feature, could you provide a concrete example of how one might use this? All of the scenarios I'm aware of where this level of debugging would be needed, I'd have access to the configuration provided in one form or another...

marwan-at-work commented:
@tnextday 

Say you are deploying Athens to a cloud provider and you want to pass specific configs. You have two options: 

1. Create your own image that extends `gomods/athesn:v0.10.0`, and add your own config file. 
2. Use Environment Variables to override some config that are either the default or were passed in your custom config file. 

Then, you realize something is not quite working. Maybe you misspelled an env var and didn't realize it

How can you see that your configs were correctly read by Athens? If you misspelled an Env Var, Athens won't know and won't tell you because it won't go through env vars and guess what you were trying to say of course. 

Asking the server for `/config` to see "the final results" of the config file (the combination of the file + env var overrides), would at least give you the peace of mind that your configuration was successfully read by Athens. 

Now, if you are deploying this to a serveless platform like Cloud Run, you can only ever access on port...so my suggestion fails here and a redacted implementation might be better. 

PS. If you don't think that's a good use case, I'm always happy to remove features instead of add them. The simpler we can keep Athens the easier it will be to maintain :v:

twexler commented:
> Now, if you are deploying this to a serveless platform like Cloud Run, you can only ever access on port...so my suggestion fails here and a redacted implementation might be better.

This is really the case I can think of, any of these "hey just deploy a docker container" things wouldn't allow more than one port to be exposed (usually). So having this operate on another port for security doesn't make a lot of sense, which leaves this as a security issue...and therefore a not particularly useful feature.

fedepaol commented:
Closing this, we may reopen in case we change our mind
